service/controller/azuremachinepool/handler/migration/azuremachinepool_test.go

```diff
@@ -0,0 +1,308 @@
+	"k8s.io/apimachinery/pkg/types"
+	capzexp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1beta1"
+
+	"github.com/giantswarm/azure-operator/v5/service/unittest"
+)
+
+func TestEnsureCreatedAzureMachinePoolIsCorrect(t *testing.T) {
+	testCases := []struct {
+		name                    string
+		oldAzureMachinePoolFile string
+		newAzureMachinePoolFile string
+		expected                expected
+	}{
+		{
+			name:                    "case 0: New AzureMachinePool references not updated",
+			oldAzureMachinePoolFile: "old_azuremachinepool.yaml",
+			newAzureMachinePoolFile: "new_azuremachinepool_references_not_updated.yaml",
+			expected: expected{
+				newAzureMachinePoolModificationsCount: 0,
+				oldAzureMachinePoolDeleted:            false,
+			},
+		},
+		{
+			name:                    "case 1: New AzureMachinePool status not updated",
+			oldAzureMachinePoolFile: "old_azuremachinepool.yaml",
+			newAzureMachinePoolFile: "new_azuremachinepool_status_not_updated.yaml",
+			expected: expected{
+				newAzureMachinePoolModificationsCount: 1,
+				oldAzureMachinePoolDeleted:            true,
+			},
+		},
+		{
+			name:                    "case 2: New AzureMachinePool fully created",
+			oldAzureMachinePoolFile: "old_azuremachinepool.yaml",
+			newAzureMachinePoolFile: "new_azuremachinepool_fully_created.yaml",
+			expected: expected{
+				newAzureMachinePoolModificationsCount: 0,
+				oldAzureMachinePoolDeleted:            true,
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			t.Log(tc.name)
+			ctx := context.Background()
+
+			// Load old AzureMachinePool
+			oldAzureMachinePool, err := loadOldAzureMachinePoolCR(tc.oldAzureMachinePoolFile)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Load new AzureMachinePool, the one which would be created by machinepoolexp/migration handler
+			newAzureMachinePool, err := loadAzureMachinePoolCR(tc.newAzureMachinePoolFile)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			fakeK8sClient := unittest.FakeK8sClient(oldAzureMachinePool, newAzureMachinePool)
+
+			// Create machinepool/migration handler
+			var resource *Resource
+			{
+				c := Config{
+					CtrlClient: fakeK8sClient.CtrlClient(),
+					Logger:     microloggertest.New(),
+				}
+				resource, err = New(c)
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			// Now run EnsureCreate for the new AzureMachinePool
+			err = resource.EnsureCreated(ctx, newAzureMachinePool)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Assert: New AzureMachinePool is correct
+			namespacedName := types.NamespacedName{
+				Namespace: newAzureMachinePool.Namespace,
+				Name:      newAzureMachinePool.Name,
+			}
+			newAzureMachinePoolAfterEnsureCreated := &capzexp.AzureMachinePool{}
+			verifyCR(t, fakeK8sClient.CtrlClient(), tc.name, newAzureMachinePoolAfterEnsureCreated, namespacedName)
+		})
+	}
+}
+
+func TestEnsureCreatedAzureMachinePoolUpdated(t *testing.T) {
```

Comment:
Test that checks if the expected number of new AzureMachinePool CR updates happened in the EnsureCreated func (in a way testing that the func is idempotent).